### PR TITLE
better kv allocate

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/pytorch/engine/engine.py [hunk 1]
```diff
@@ -261,6 +262,20 @@ def _response(self,
                      data=data,
                      err_msg=err_msg))
 
+    def _get_max_session_len(self):
+        """get max session len."""
+        session_len = self.scheduler_config.max_session_len
+        max_tokens = (self.cache_config.num_gpu_blocks *
+                      self.cache_config.block_size)
+        window_size = self.cache_config.window_size
+        if window_size > 0 and window_size <= max_tokens:
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/engine.py:271]: what's the `window_size`?

Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:271]: sliding window attention

Comment (lvhan028) [reply lmdeploy/pytorch/engine/engine.py:271]: Suppose `window_size == 2048` and `session_len is None`
According to the logic, `session_len = max_tokens = (1 << 63) - 1`
Is it expected?

**Review by AllentDan — APPROVED** [review]
LGTM

File: lmdeploy/pytorch/engine/engine.py (hunk 1 above)
Comment (AllentDan) [reply lmdeploy/pytorch/engine/engine.py:271]: Does it mean that sliding window attention supports enumerous session length?

Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:271]: > Does it mean that sliding window attention supports infinite session length?
Yes

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**